core/src/worker/tuner/resource_based.rs

```diff
@@ -495,26 +497,40 @@ struct RealSysInfoInner {
     total_mem: AtomicU64,
     cur_mem_usage: AtomicU64,
     cur_cpu_usage: AtomicU64,
+    cgroup_cpu_info: CGroupCpuInfo<CgroupV2CpuFileSystem>,
 }
 
 impl RealSysInfoInner {
     fn refresh(&self) {
         let mut lock = self.sys.lock();
         lock.refresh_memory();
-        lock.refresh_cpu_usage();
-        let cpu = lock.global_cpu_usage() as f64 / 100.;
         if let Some(cgroup_limits) = lock.cgroup_limits() {
             self.total_mem
                 .store(cgroup_limits.total_memory, Ordering::Release);
             self.cur_mem_usage.store(
                 cgroup_limits.total_memory - cgroup_limits.free_memory,
                 Ordering::Release,
             );
+
+            let cpu = self.cgroup_cpu_info.calc_cpu_percent().unwrap_or_else(|| {
+                //there won't be a cgroup cpu usage if there is no limit applied to the cgroup
+                //or if an error is encountered when reading cpu.stat or cpu.max
+                //in these cases, fallback to global cpu usage
```

Comment:
Nit: I'm vaguely surprised this isn't a cargo fmt thing, but, the lack a space after your `//` freaks me out